proto-grpc.md

```diff
@@ -0,0 +1,277 @@
+reflect today's state.
+
+
+## The Basics
+
+At its most basic level, the task of working with Protobuf and gRPC `.proto` files is much the same
+as outside of Bazel: we execute the Protobuf Compiler `protoc` with the necessary arguments,
+environment and plugins to get it to produce a desired set of files or libraries for any specific
+language. For those unfamiliar with protoc, it is a small compiler executable that parses sets of
+`.proto` formatted files and executes plugins to write a set of output source code files. These
+files implement the interfaces for working with the messages and services in the language supported
+by the plugin.
+
+However, given the prevalence of Protobuf and gRPC use, users are not expected to reinvent the rules
+needed for calling protoc themselves. Instead, the common base dependency ruleset
+[bazelbuild/rules_proto](https://github.com/bazelbuild/rules_proto) provides the `proto_library`
+rule which can be used to establish libraries of `.proto` files and map their dependencies. For
+example, suppose we have the following dependency tree of `.proto` files:
+
+![Tree of proto files](./assets/proto-tree.png)
+
+For this set of files, we could define the following collection of `proto_library` targets. Note
+that some rulesets introduced later, such as stackb/rules_proto and rules_go, provide tooling
+through Gazelle for auto-generating these `proto_library` from the `.proto` files.
+
+```starlark
+load("@rules_proto//proto:defs.bzl", "proto_library")
+
+proto_library(
+  name = "proto_a",
+  srcs = ["A.proto"],
+)
+
+proto_library(
+  name = "proto_b",
+  srcs = ["B.proto"],
+)
+
+proto_library(
+  name = "proto_c",
+  srcs = ["C.proto"],
+  deps = [":proto_a", ":proto_b"], # Provides the depended files when calling protoc
+)
+```
+
+Using `bazel query`, we can view the graph that Bazel has assembled from the targets, showing it
+correctly reflects the dependency of `proto_c` on `proto_a` and `proto_b` (you must have
+DOT/GraphViz installed):
+
+`bazel query --output graph --nograph:factored 'kind(proto_library, //...)'  | dot -Tpng > graph.png`
+
+![Generated graph of proto_library targets](./assets/proto-bazel-query.png)
+
+
+Note that we could have grouped all of these `.proto` files under a single `proto_library` in this
+particular case. In real codebases these files may be under separate directories and managed by
+different groups or companies. The convention for splitting or grouping libraries varies by target
+language, but for most it is a matter of local preference. However, for some languages, files with
+the same internal [`package`](https://developers.google.com/protocol-buffers/docs/proto3#packages)
+specifier should normally be grouped into a single library. In general, you should be
+consistent with your grouping conventions within your workspace.
+
+At this point we have defined the abstract collections of `.proto` files, but these targets alone
```

Comment:
Maybe mention that `.proto` files are checked for correctness (syntax, imports, ...) when building `proto_library` targets?

Reply:
I've added this now where I introduce proto_library